# MLflow results presentation (2)

In [1]:
from functools import lru_cache
import pandas as pd
from uncertainties import ufloat
import mlflow
from mlflow.tracking import MlflowClient
mlflow.set_tracking_uri('file:///home/zeyusun/work/flare-prediction-smarp/mlruns')
client = MlflowClient()

from mlflow_helper import *
retrieve = lru_cache(retrieve)

pd.set_option('display.max_colwidth', None)
pd.set_option('display.max_rows', 500)

%load_ext tensorboard
%load_ext autoreload
%autoreload 2

In [2]:
%%HTML
<style type="text/css">
table.dataframe td, table.dataframe th {
    border: 1px  black solid !important;
  color: black !important;
}
</style>

## Compare leaderboard 3: LSTM_CNN vs base

In [ ]:
def compare(base_run_name, target_run_name):
    base_exp_name, target_exp_name = ['leaderboard3'] * 2
    if '/' in base_run_name:
        base_exp_name, base_run_name = base_run_name.split('/')
    if '/' in target_run:
        target_exp_name, target_run_name = target_run_name.split('/')

    runs_raw_1 = retrieve('leaderboard3', 'LSTM_CNN')
    runs_1 = select(runs_raw_1)
    organizes

In [81]:
runs_raw_1 = retrieve('leaderboard3', 'LSTM_CNN')
runs_1 = select(runs_raw_1)

ckpt = runs_raw_1['tags.checkpoint'].str.extract(r'epoch=(?P<epoch>[0-9]+)-step=(?P<step>[0-9]+)').astype(int)
runs_1 = pd.concat((runs_1, ckpt), axis=1)

runs_1.style.background_gradient(axis=0)

Select iloc 0 from 
                          start_time tags.mlflow.runName  \
207 2021-06-27 16:03:33.239000+00:00            LSTM_CNN   

                tags.mlflow.source.git.commit  
207  e13810ffddafdb818df7c4de68d9eafd67861e8c  


In [83]:
organize(runs_1, std=True)

dataset                  fused_sharp          fused_smarp  \
      estimator                                             
ACC   CNN              0.943+/-0.030        0.910+/-0.019   
      LSTM             0.957+/-0.012        0.921+/-0.016   
AUC   CNN              0.983+/-0.013        0.964+/-0.012   
      LSTM             0.988+/-0.004        0.970+/-0.010   
TSS   CNN              0.886+/-0.059        0.820+/-0.038   
      LSTM             0.913+/-0.025        0.841+/-0.032   
HSS   CNN              0.886+/-0.059        0.820+/-0.038   
      LSTM             0.913+/-0.025        0.841+/-0.032   
BSS   CNN              0.343+/-0.388        0.365+/-0.305   
      LSTM             0.793+/-0.055        0.621+/-0.056   
epoch CNN              6.600+/-4.980        9.200+/-7.530   
      LSTM            10.600+/-7.537       16.200+/-2.588   
step  CNN         1301.600+/-878.957  1582.200+/-1135.646   
      LSTM       1933.800+/-1228.537   2716.000+/-436.486   

dataset                        sharp               smarp  
      estimator                                           
ACC   CNN              0.934+/-0.011       0.888+/-0.021  
      LSTM             0.951+/-0.012       0.918+/-0.017  
AUC   CNN              0.981+/-0.004       0.956+/-0.016  
      LSTM             0.986+/-0.005       0.965+/-0.009  
TSS   CNN              0.867+/-0.023       0.776+/-0.041  
      LSTM             0.901+/-0.024       0.836+/-0.034  
HSS   CNN              0.867+/-0.023       0.776+/-0.041  
      LSTM             0.901+/-0.024       0.836+/-0.034  
BSS   CNN              0.619+/-0.125       0.274+/-0.493  
      LSTM             0.757+/-0.062       0.581+/-0.047  
epoch CNN            28.000+/-19.545     16.000+/-10.223  
      LSTM             8.200+/-7.190      17.000+/-1.225  
step  CNN        1668.800+/-1263.607  1567.200+/-962.466  
      LSTM         491.600+/-356.530  1664.400+/-109.703

In [84]:
runs_raw_2 = retrieve('leaderboard3', 'base')
runs_2 = select(runs_raw_2)

ckpt = runs_raw_2['tags.checkpoint'].str.extract(r'epoch=(?P<epoch>[0-9]+)-step=(?P<step>[0-9]+)').astype(int)
runs_2 = pd.concat((runs_2, ckpt), axis=1)

runs_2.style.background_gradient(axis=0)

Select iloc 0 from 
                          start_time tags.mlflow.runName  \
124 2021-08-18 03:00:51.425000+00:00                base   

                tags.mlflow.source.git.commit  
124  98d231a7d80767545d12d384c881f6eb545ffb7a  


In [85]:
organize(runs_2, std=True)

dataset                 fused_sharp         fused_smarp              sharp  \
      estimator                                                              
ACC   CNN             0.892+/-0.076       0.839+/-0.072      0.772+/-0.081   
      LSTM            0.944+/-0.012       0.909+/-0.026      0.934+/-0.023   
AUC   CNN             0.980+/-0.010       0.968+/-0.006      0.967+/-0.029   
      LSTM            0.990+/-0.005       0.971+/-0.010      0.985+/-0.004   
TSS   CNN             0.784+/-0.152       0.678+/-0.143      0.545+/-0.162   
      LSTM            0.888+/-0.023       0.818+/-0.052      0.867+/-0.046   
HSS   CNN             0.784+/-0.152       0.678+/-0.143      0.545+/-0.162   
      LSTM            0.888+/-0.023       0.818+/-0.052      0.867+/-0.046   
BSS   CNN             0.569+/-0.359       0.376+/-0.286      0.101+/-0.346   
      LSTM            0.797+/-0.052       0.642+/-0.103      0.736+/-0.076   
epoch CNN             1.400+/-1.517       2.600+/-2.191     10.200+/-4.712   
      LSTM            5.200+/-3.421       5.600+/-3.362      3.600+/-2.881   
step  CNN         407.000+/-263.260   569.800+/-344.752  608.800+/-196.895   
      LSTM       1041.400+/-557.124  1022.000+/-482.548  250.400+/-153.295   

dataset                      smarp  
      estimator                     
ACC   CNN            0.879+/-0.014  
      LSTM           0.911+/-0.015  
AUC   CNN            0.958+/-0.014  
      LSTM           0.970+/-0.009  
TSS   CNN            0.758+/-0.028  
      LSTM           0.822+/-0.030  
HSS   CNN            0.758+/-0.028  
      LSTM           0.822+/-0.030  
BSS   CNN            0.544+/-0.049  
      LSTM           0.629+/-0.049  
epoch CNN            6.400+/-3.847  
      LSTM           7.400+/-1.817  
step  CNN        689.200+/-368.656  
      LSTM       780.400+/-187.720

#### Sharp problem
* What: In `base` run, CNN trained on SHARP withseed 3 has particularly bad performance.
* Why: AUC is a bad metric. The step 494 (end of epoch 8) has higher validation AUC than its neighboring epochs (not very distinguishable, 0.994 vs 0.986 and 0.989), but other metrics are lower (e.g., 0.702 vs 0.840 and 0.883). Looking at SSP, the curve significantly tilted leftwards, indicating a significant amount of underforecasting.
* This also suggests using an optimized TSS is a dangerous idea because it can select a under- or over-forecasting model that is not robust to the threshold.
    * It is very much possible that the test SSP is a bit different from the validation SSP, and using the threshold fitted on the validation set may lead to a performance much lower than expected. This is due to the sensitivity of SSP to the threshold.
    * Thresholding on the test set may not be very useful in practice, as we are not allowed to give a prediction of a test instance unless we know the entire test set.
* Why: AUC is not consistent with others is because it is the area under the uniform measure of false alarm rate, which is rarely the case in practice where false alarm rate is selected to be close to zero. AUC gives too much weight on the part where false alarm is high.
    * Area Under SSP may be a good idea. It is the area using the uniform measure of threshold.

In [88]:
runs_comb = pd.concat((
    runs_raw_1.loc[[198]],
    runs_raw_2.loc[[115]],
))
runs_comb

run_id experiment_id    status  \
198  f7bc8479849e4a30805eb7cfd116395e            15  FINISHED   
115  dc8a1de8327b49838879eb7ec444feed            15  FINISHED   

                                                                                              artifact_uri  \
198  file:///home/zeyusun/work/flare-prediction-smarp/mlruns/15/f7bc8479849e4a30805eb7cfd116395e/artifacts   
115  file:///home/zeyusun/work/flare-prediction-smarp/mlruns/15/dc8a1de8327b49838879eb7ec444feed/artifacts   

                          start_time                         end_time  \
198 2021-06-27 16:49:13.611000+00:00 2021-06-27 17:03:02.065000+00:00   
115 2021-08-18 03:39:51.850000+00:00 2021-08-18 03:48:29.301000+00:00   

     metrics.test/auc  metrics.test/recall  metrics.validation/accuracy  \
198          0.979329             0.953096                     0.926554   
115          0.916192             0.307692                     0.860640   

     metrics.test/hss1  ...  tags.mlflow.source.name  \
198           0.853659  ...           sweep_arnet.py   
115           0.300188  ...           sweep_arnet.py   

              tags.mlflow.parentRunId  tags.mlflow.source.type  tags.balanced  \
198  0d9e66f8c3ca4266a4ca374e4d3b377c                    LOCAL           True   
115  b168fac075e243be9bc28f888ef21a23                    LOCAL           True   

     tags.mlflow.user  tags.database_name  \
198           zeyusun            M_Q_24hr   
115           zeyusun            M_Q_24hr   

                tags.mlflow.source.git.commit  tags.estimator_name  \
198  e13810ffddafdb818df7c4de68d9eafd67861e8c                  CNN   
115  98d231a7d80767545d12d384c881f6eb545ffb7a                  CNN   

                                                                                              tags.checkpoint  \
198  /home/zeyusun/work/flare-prediction-smarp/lightning_logs/version_953/checkpoints/epoch=34-step=1924.ckpt   
115   /home/zeyusun/work/flare-prediction-smarp/lightning_logs/version_1275/checkpoints/epoch=8-step=494.ckpt   

     tags.mlflow.runName  
198   M_Q_24hr_CNN_sharp  
115   M_Q_24hr_CNN_sharp  

[2 rows x 81 columns]

In [89]:
dirs = tensorboard(runs_comb)
%tensorboard --logdir_spec {dirs}

Reusing TensorBoard on port 6008 (pid 32279), started 8:22:49 ago. (Use '!kill 32279' to kill it.)

#### Fused_smarp problem
* What: fused_smarp is worse than smarp in all metrics except for AUC. LSTM_CNN has a similar AUC to base, but other metrics are much better than base.
* Why: We think maybe it's because base has threahold 0.5 is the problem.
    * In `base` run, CNN trained on fused_smarp with seed 2 is a perfect piece of evidence that AUC is a bad validation metric. According to AUC, epoch 2 has is the best. But SSP is severely tilted. Let's use TSS.

In [86]:
runs_comb = pd.concat((
    runs_raw_1.loc[[170]],
    runs_raw_2.loc[[87]],
))
runs_comb

run_id experiment_id    status  \
170  ea083959ee884688966dbaeac8079b08            15  FINISHED   
87   fabff25c06b34516b1b229b6c45ca1b4            15  FINISHED   

                                                                                              artifact_uri  \
170  file:///home/zeyusun/work/flare-prediction-smarp/mlruns/15/ea083959ee884688966dbaeac8079b08/artifacts   
87   file:///home/zeyusun/work/flare-prediction-smarp/mlruns/15/fabff25c06b34516b1b229b6c45ca1b4/artifacts   

                          start_time                         end_time  \
170 2021-06-27 21:00:49.510000+00:00 2021-06-27 21:21:07.944000+00:00   
87  2021-08-18 07:28:38.833000+00:00 2021-08-18 07:42:04.967000+00:00   

     metrics.test/auc  metrics.test/recall  metrics.validation/accuracy  \
170          0.968733             0.892280                     0.864382   
87           0.967358             0.473968                     0.784749   

     metrics.test/hss1  ...  tags.mlflow.source.name  \
170           0.825853  ...           sweep_arnet.py   
87            0.464093  ...           sweep_arnet.py   

              tags.mlflow.parentRunId  tags.mlflow.source.type  tags.balanced  \
170  0d9e66f8c3ca4266a4ca374e4d3b377c                    LOCAL           True   
87   b168fac075e243be9bc28f888ef21a23                    LOCAL           True   

     tags.mlflow.user  tags.database_name  \
170           zeyusun            M_Q_24hr   
87            zeyusun            M_Q_24hr   

                tags.mlflow.source.git.commit  tags.estimator_name  \
170  e13810ffddafdb818df7c4de68d9eafd67861e8c                  CNN   
87   98d231a7d80767545d12d384c881f6eb545ffb7a                  CNN   

                                                                                             tags.checkpoint  \
170  /home/zeyusun/work/flare-prediction-smarp/lightning_logs/version_981/checkpoints/epoch=9-step=1649.ckpt   
87   /home/zeyusun/work/flare-prediction-smarp/lightning_logs/version_1303/checkpoints/epoch=2-step=494.ckpt   

          tags.mlflow.runName  
170  M_Q_24hr_CNN_fused_smarp  
87   M_Q_24hr_CNN_fused_smarp  

[2 rows x 81 columns]

In [87]:
dirs = tensorboard(runs_comb)
%tensorboard --logdir_spec {dirs}

In [64]:
runs_diff = diff(runs_1[runs_1['estimator'] == 'CNN'], runs_2[runs_2['estimator'] == 'CNN'])
runs_diff.style.background_gradient(axis=0, vmin=-0.2, vmax=0.2, cmap='seismic')

## Compare leaderboard3: base vs base_rep_CNN
CNN on fused_sharp with seed 2 also proves AUC is a bad metric. High AUC may well associates with a low TSS in the under- or over-forecasting (if the threshold is 0.5)

In [90]:
runs_raw_1 = retrieve('leaderboard3', 'base')
runs_1 = select(runs_raw_1)
runs_1.style.background_gradient(axis=0)

Select iloc 0 from 
                          start_time tags.mlflow.runName  \
124 2021-08-18 03:00:51.425000+00:00                base   

                tags.mlflow.source.git.commit  
124  98d231a7d80767545d12d384c881f6eb545ffb7a  


In [91]:
runs_raw_2 = retrieve('leaderboard3', 'base_rep_CNN')
runs_2 = select(runs_raw_2)
runs_2.style.background_gradient(axis=0)

Select iloc 0 from 
                         start_time tags.mlflow.runName  \
42 2021-08-19 05:07:03.892000+00:00        base_rep_CNN   

               tags.mlflow.source.git.commit  
42  ed2b596b684b6bd9d96d47271a0e01b39da971d2  


In [92]:
runs_diff = diff(runs_1[runs_1['estimator'] == 'CNN'], runs_2)
runs_diff.style.background_gradient(axis=0, vmin=-0.2, vmax=0.2, cmap='seismic')

In [93]:
runs_raw_3 = retrieve('leaderboard3', 'val_same_test')
runs_3 = select(runs_raw_3)
runs_3.style.background_gradient(axis=0)

Select iloc 0 from 
                         start_time tags.mlflow.runName  \
83 2021-08-18 21:35:53.919000+00:00       val_same_test   

               tags.mlflow.source.git.commit  
83  717ab7690cf682989b9b190c2a8653f09ea4fdff  


In [94]:
runs_diff = diff(runs_1[runs_1['estimator'] == 'CNN'], runs_3[runs_3['estimator'] == 'CNN'])
runs_diff.style.background_gradient(axis=0, vmin=-0.2, vmax=0.2, cmap='seismic')

In [95]:
runs_diff = diff(runs_2, runs_3[runs_3['estimator'] == 'CNN'])
runs_diff.style.background_gradient(axis=0, vmin=-0.2, vmax=0.2, cmap='seismic')

## Compare leaderboard3: base_rep_CNN vs larger_lr
`Does larger learning rate help?`
* It doesn't even make sense to compare the performance if you don't select the robust model.
* It seems larger learning rate causes, though almost indistinguishable, slower convergence of the training loss (fused_sharp with seed 3 and 4, sharp with seed 3), which is interesting.

In [96]:
runs_raw_1 = retrieve('leaderboard3', 'base_rep_CNN')
runs_1 = select(runs_raw_1)
runs_1.style.background_gradient(axis=0)

Select iloc 0 from 
                         start_time tags.mlflow.runName  \
42 2021-08-19 05:07:03.892000+00:00        base_rep_CNN   

               tags.mlflow.source.git.commit  
42  ed2b596b684b6bd9d96d47271a0e01b39da971d2  


In [97]:
runs_raw_2 = retrieve('leaderboard3', 'larger_lr')
runs_2 = select(runs_raw_2)
runs_2.style.background_gradient(axis=0)

Select iloc 0 from 
                         start_time tags.mlflow.runName  \
27 2021-08-19 08:03:56.682000+00:00           larger_lr   

               tags.mlflow.source.git.commit  
27  10eedfd9086bcc850e8538fa1ca9c6975495c544  


In [98]:
runs_diff = diff(runs_1[runs_1['dataset'].isin(['fused_sharp', 'sharp'])], runs_2)
runs_diff.style.background_gradient(axis=0, vmin=-0.2, vmax=0.2, cmap='seismic')

In [103]:
runs_comb = pd.concat((
    runs_raw_1.loc[[38]],
    runs_raw_2.loc[[15]],
))
runs_comb

run_id experiment_id    status  \
38  518b518563574a2b9a01a51bbe040ad9            15  FINISHED   
15  9d16191e1e84476dbeb2ded425b323ed            15  FINISHED   

                                                                                             artifact_uri  \
38  file:///home/zeyusun/work/flare-prediction-smarp/mlruns/15/518b518563574a2b9a01a51bbe040ad9/artifacts   
15  file:///home/zeyusun/work/flare-prediction-smarp/mlruns/15/9d16191e1e84476dbeb2ded425b323ed/artifacts   

                         start_time                         end_time  \
38 2021-08-19 05:35:05.410000+00:00 2021-08-19 05:45:24.838000+00:00   
15 2021-08-19 11:00:01.676000+00:00 2021-08-19 12:48:34.576000+00:00   

    metrics.test/auc  metrics.test/recall  metrics.validation/accuracy  \
38          0.970089             0.789869                     0.908663   
15          0.970372             0.744841                     0.838041   

    metrics.test/hss1  ...  tags.mlflow.source.name  \
38           0.754221  ...           sweep_arnet.py   
15           0.729831  ...           sweep_arnet.py   

             tags.mlflow.parentRunId  tags.mlflow.source.type  tags.balanced  \
38  6f8716d1ce714738b3f524a87c37512b                    LOCAL           True   
15  e6062e9ba5354837846d0a656d39044f                    LOCAL           True   

    tags.mlflow.user  tags.database_name  \
38           zeyusun            M_Q_24hr   
15           zeyusun            M_Q_24hr   

               tags.mlflow.source.git.commit  tags.estimator_name  \
38  ed2b596b684b6bd9d96d47271a0e01b39da971d2                  CNN   
15  10eedfd9086bcc850e8538fa1ca9c6975495c544                  CNN   

                                                                                             tags.checkpoint  \
38   /home/zeyusun/work/flare-prediction-smarp/lightning_logs/version_1350/checkpoints/epoch=7-step=439.ckpt   
15  /home/zeyusun/work/flare-prediction-smarp/lightning_logs/version_1371/checkpoints/epoch=15-step=879.ckpt   

    tags.mlflow.runName  
38   M_Q_24hr_CNN_sharp  
15   M_Q_24hr_CNN_sharp  

[2 rows x 81 columns]

In [105]:
dirs = tensorboard(runs_comb)
%tensorboard --logdir_spec {dirs}

## Compare leaderboard3: cnn_li2020

In [47]:
runs_raw_2 = retrieve('leaderboard3', 'cnn_li2020')
runs_2 = select(runs_raw_2)
runs_2.style.background_gradient(axis=0)

Select iloc 0 from 
                         start_time tags.mlflow.runName  \
22 2021-08-19 09:04:34.552000+00:00          cnn_li2020   

               tags.mlflow.source.git.commit  
22  167b6d8cde9a134dba8cfe12a142cc97c97dcb58  


In [48]:
runs_raw_1 = retrieve('leaderboard3', 'base_rep_CNN')
runs_raw_1 = runs_raw_1[runs_raw_1['tags.dataset_name'].isin(['fused_sharp', 'sharp'])]
runs_1 = select(runs_raw_1)
runs_1.style.background_gradient(axis=0)

Select iloc 0 from 
                         start_time tags.mlflow.runName  \
42 2021-08-19 05:07:03.892000+00:00        base_rep_CNN   

               tags.mlflow.source.git.commit  
42  ed2b596b684b6bd9d96d47271a0e01b39da971d2  


In [49]:
runs_diff = diff(runs_1, runs_2)
runs_diff.style.background_gradient(axis=0, vmin=-0.2, vmax=0.2, cmap='seismic')

* Code difference
|                  |   1: base_rep_CNN   |   2: cnn_li2020   |
|:----------------:|:-------------------:|:-----------------:|
|     patience     |          10         |         20        |
|   learning rate  |         1e-3        |        1e-2       |
| 1/5 of the other | used for validation | used for training |
* Performance difference
    * base_rep_CNN is very bad: fused_sharp's performance is all significantly worse than sharp's except for seed 1.
    * cnn_li2020: fused_sharp sees an increase in ACC but not AUC

## two_stages:test

In [56]:
runs_raw = retrieve('two_stages', 'test')
runs = select(runs_raw)
runs.style.background_gradient(axis=0)

Select iloc 0 from 
                         start_time tags.mlflow.runName  \
1  2021-08-22 06:36:45.309000+00:00                test   
3  2021-08-22 06:25:06.680000+00:00                test   
5  2021-08-22 06:22:07.482000+00:00                test   
8  2021-08-21 08:33:47.985000+00:00                test   
10 2021-08-21 08:28:55.500000+00:00                test   

               tags.mlflow.source.git.commit  
1   48c5794285044d9e7fe19bc0cb0238df3892da66  
3   48c5794285044d9e7fe19bc0cb0238df3892da66  
5   48c5794285044d9e7fe19bc0cb0238df3892da66  
8   1d5709bbc561def0dab992f1cd1f3d1a0acfdaa1  
10  1d5709bbc561def0dab992f1cd1f3d1a0acfdaa1  


In [57]:
dirs = tensorboard(runs_raw)
%tensorboard --logdir_spec {dirs}

## leaderboard3:val_tss

* LSTM: compare sharp vs fused_sharp
    * Seed 0 in mlflow using the [mlflow query](http://localhost:5000/#/metric/train%2Floss?runs=[%22545a671d3fe643b185a788e9ec5ef2dc%22,%22a149845829bc4bec9e4028a531313ddf%22]&experiment=15&plot_metric_keys=[%22train/loss%22,%22validation0/tss%22,%22validation1/tss%22,%22train/epoch%22,%22test/tss%22]&plot_layout={%22autosize%22:true,%22xaxis%22:{%22autorange%22:true,%22type%22:%22linear%22},%22yaxis%22:{%22range%22:[0.7144796062068293,1.063855720288826],%22autorange%22:false,%22type%22:%22linear%22}}&x_axis=step&y_axis_scale=linear&line_smoothness=1&show_point=true&deselected_curves=[]&last_linear_y_axis_range=[]), we find that
        * Fused_sharp's tss plateaus higher than sharp.
        * There is some discrepancy between val and test, which results in validation not selecting the best epoch.
    * Seed 1 using [query](http://localhost:5000/#/metric/train%2Floss?runs=[%228077ad3678e2449ca4d71aa3986a3c82%22,%2210b430d66c824ddbaf834fe7476ebcf2%22]&experiment=15&plot_metric_keys=[%22train/loss%22,%22validation0/tss%22,%22validation1/tss%22,%22train/epoch%22,%22test/tss%22]&plot_layout={%22autosize%22:true,%22xaxis%22:{%22autorange%22:true,%22type%22:%22linear%22},%22yaxis%22:{%22range%22:[0.792408096599069,0.9987583639787481],%22autorange%22:false,%22type%22:%22linear%22}}&x_axis=step&y_axis_scale=linear&line_smoothness=1&show_point=true&deselected_curves=[]&last_linear_y_axis_range=[])
        * Similar to seed 0. Fused_sharp outperformes.
    * Seed 2 using [query]()
        * Fused_sharp test/tss is better
    * Seed 3 using [query](http://localhost:5000/#/metric/train%2Floss?runs=[%2250fe2d2c83a04f46abf592a4728afd6d%22,%22e429f17cce5a4e79b90a285e8bc17928%22]&experiment=15&plot_metric_keys=[%22train/loss%22,%22validation0/tss%22,%22validation1/tss%22,%22train/epoch%22,%22test/tss%22]&plot_layout={%22autosize%22:true,%22xaxis%22:{%22autorange%22:true,%22type%22:%22linear%22},%22yaxis%22:{%22range%22:[0.7477693999727355,0.9750198456055341],%22autorange%22:false,%22type%22:%22linear%22}}&x_axis=step&y_axis_scale=linear&line_smoothness=1&show_point=true&deselected_curves=[]&last_linear_y_axis_range=[])
        * Fused_sharp test/tss is worse
        * Actually Fused_sharp has test tss curve above sharp. But the validation selects a bad initial epoch which has the top validation tss but not so much for the test set. It has very much the potential to outperform, just with a little more patience and more candidates. This is a sign to increase the patience to from 5 to 10. More patience doesn't hurt because validation performance is a good estimator for the test.
    * Seed 4 using [query](http://localhost:5000/#/metric/train%2Floss?runs=[%226fc1fedb68644766a4da2d2d31e73f29%22,%2290e6cf9fa69449a1b3e68cbc2f2ff772%22]&experiment=15&plot_metric_keys=[%22train/loss%22,%22validation0/tss%22,%22validation1/tss%22,%22train/epoch%22,%22test/tss%22]&plot_layout={%22autosize%22:true,%22xaxis%22:{%22autorange%22:true,%22type%22:%22linear%22},%22yaxis%22:{%22range%22:[0.8171573726952396,0.9319397089251861],%22autorange%22:false,%22type%22:%22linear%22}}&x_axis=step&y_axis_scale=linear&line_smoothness=1&show_point=true&deselected_curves=[]&last_linear_y_axis_range=[])
        * Just another example of Seed 3. Fused_sharp has much fewer epochs than sharp
* CNN: compare sharp vs fused_sharp
    * Seed 0 with [query](http://localhost:5000/#/metric/train%2Floss?runs=[%22d06fc4fcb32f4ebebbf2f1739a7cdab4%22,%2299519d6879da4ef6b1d1156bd901fbe1%22]&experiment=15&plot_metric_keys=[%22train/loss%22,%22validation0/tss%22,%22validation1/tss%22,%22train/epoch%22,%22test/tss%22]&plot_layout={%22autosize%22:true,%22xaxis%22:{%22autorange%22:false,%22type%22:%22linear%22,%22range%22:[-31.93973827362052,3979.3755865851376]},%22yaxis%22:{%22range%22:[0.45905119986032605,0.9746929110902727],%22autorange%22:false,%22type%22:%22linear%22}}&x_axis=step&y_axis_scale=linear&line_smoothness=1&show_point=false&deselected_curves=[]&last_linear_y_axis_range=[])
        * Fused_sharp test/tss is worse.
        * Fused_sharp has validation curve above sharp's validation curve. But sharp has test curve above its validation and it syncs well with its validation. On the other hand, fused_sharp's test curve goes a bit down below its validation curve as training goes on. It doesn't seem to be a good approximation and a very validation tss only corresponds to a mediocre test tss, which is lower than sharp's.
            * The validation0 and validation1 of sharp and fused_sharp agree.
        * It is still possible to perform well on the validation set and badly on the test set.
        * Seen from the reliability diagram, differences between epochs at the end of training is still large.
    * Seed 1 with [query](http://localhost:5000/#/metric/train%2Floss?runs=[%22daad85e9fd0249b5ad07d7063d5e830d%22,%225d20fe5e095341eab9e6af192f42d296%22]&experiment=15&plot_metric_keys=[%22train/loss%22,%22validation0/tss%22,%22validation1/tss%22,%22train/epoch%22,%22test/tss%22]&plot_layout={%22autosize%22:true,%22xaxis%22:{%22autorange%22:true,%22type%22:%22linear%22,%22range%22:[-31.93973827362052,3979.3755865851376]},%22yaxis%22:{%22range%22:[0.2844251336898397,1.0544786096256686],%22autorange%22:false,%22type%22:%22linear%22}}&x_axis=step&y_axis_scale=linear&line_smoothness=1&show_point=false&deselected_curves=[]&last_linear_y_axis_range=[])
        * Fused_sharp test/tss is worse.
        * Fused_sharp doesn't even have validation curve above sharp's validation curve.
    * Seed 2 with [query](http://localhost:5000/#/metric/train%2Floss?runs=[%226464c469739b45e1bd4f40192c87eadb%22,%2254b434401ef24513b17a15fc13534590%22]&experiment=15&plot_metric_keys=[%22train/loss%22,%22validation0/tss%22,%22validation1/tss%22,%22train/epoch%22,%22test/tss%22]&plot_layout={%22autosize%22:true,%22xaxis%22:{%22autorange%22:true,%22type%22:%22linear%22,%22range%22:[-31.93973827362052,3979.3755865851376]},%22yaxis%22:{%22range%22:[0.2844251336898397,1.0544786096256686],%22autorange%22:false,%22type%22:%22linear%22}}&x_axis=step&y_axis_scale=linear&line_smoothness=1&show_point=false&deselected_curves=[]&last_linear_y_axis_range=[])

In [6]:
runs_raw_1 = retrieve('leaderboard3', 'val_tss')
runs_1 = select(runs_raw_1)
runs_1.style.background_gradient(axis=0)

Select iloc slice(None, None, None) from 
                          start_time tags.mlflow.runName
245 2021-08-23 09:22:37.321000+00:00             val_tss


In [30]:
for estimator in ['LSTM', 'CNN']:
    for dataset in ['fused_sharp', 'sharp', 'fused_smarp', 'smarp']:
        download_figures(runs_raw_1, dataset, 1, estimator)

In [35]:
runs_1.loc[runs_1['estimator'] == 'LSTM'].style.background_gradient(axis=0)

In [4]:
organize(runs_1, std=True)

dataset          fused_sharp    fused_smarp          sharp          smarp
    estimator                                                            
ACC CNN        0.897+/-0.047  0.916+/-0.008  0.914+/-0.016  0.883+/-0.024
    LSTM       0.950+/-0.013  0.913+/-0.011  0.936+/-0.018  0.909+/-0.013
AUC CNN        0.978+/-0.011  0.972+/-0.003  0.980+/-0.007  0.955+/-0.014
    LSTM       0.988+/-0.004  0.970+/-0.010  0.985+/-0.004  0.968+/-0.011
TSS CNN        0.793+/-0.095  0.831+/-0.016  0.828+/-0.033  0.766+/-0.048
    LSTM       0.899+/-0.026  0.827+/-0.023  0.873+/-0.036  0.818+/-0.026
HSS CNN        0.793+/-0.095  0.831+/-0.016  0.828+/-0.033  0.766+/-0.048
    LSTM       0.899+/-0.026  0.827+/-0.023  0.873+/-0.036  0.818+/-0.026
BSS CNN        0.607+/-0.197  0.692+/-0.036  0.690+/-0.069  0.555+/-0.097
    LSTM       0.794+/-0.041  0.662+/-0.057  0.755+/-0.045  0.626+/-0.072

In [7]:
print(typeset(organize(runs_1, std=True)))

\begin{tabular}{llllll}
\toprule
    & dataset &    fused\_sharp &    fused\_smarp &          sharp &          smarp \\
{} & estimator &                &                &                &                \\
\midrule
ACC & CNN &  0.897+/-0.047 &  0.916+/-0.008 &  0.914+/-0.016 &  0.883+/-0.024 \\
    & LSTM &  0.950+/-0.013 &  0.913+/-0.011 &  0.936+/-0.018 &  0.909+/-0.013 \\
AUC & CNN &  0.978+/-0.011 &  0.972+/-0.003 &  0.980+/-0.007 &  0.955+/-0.014 \\
    & LSTM &  0.988+/-0.004 &  0.970+/-0.010 &  0.985+/-0.004 &  0.968+/-0.011 \\
TSS & CNN &  0.793+/-0.095 &  0.831+/-0.016 &  0.828+/-0.033 &  0.766+/-0.048 \\
    & LSTM &  0.899+/-0.026 &  0.827+/-0.023 &  0.873+/-0.036 &  0.818+/-0.026 \\
HSS & CNN &  0.793+/-0.095 &  0.831+/-0.016 &  0.828+/-0.033 &  0.766+/-0.048 \\
    & LSTM &  0.899+/-0.026 &  0.827+/-0.023 &  0.873+/-0.036 &  0.818+/-0.026 \\
BSS & CNN &  0.607+/-0.197 &  0.692+/-0.036 &  0.690+/-0.069 &  0.555+/-0.097 \\
    & LSTM &  0.794+/-0.041 &  0.662+/-0.057 &  0.75

In [10]:
df_ttest = tabulate_pvalues(runs_1)

In [11]:
df_ttest

S estimator          tested hypothesis         t   p-value
0   ACC      LSTM  S(fused_sharp) > S(sharp)  2.691665  0.027282
1   AUC      LSTM  S(fused_sharp) > S(sharp)  2.669280  0.027924
2   TSS      LSTM  S(fused_sharp) > S(sharp)  2.691671  0.027282
3   HSS      LSTM  S(fused_sharp) > S(sharp)  2.691672  0.027282
4   BSS      LSTM  S(fused_sharp) > S(sharp)  2.351116  0.039210
5   ACC       CNN  S(fused_sharp) > S(sharp) -0.821323  0.771203
6   AUC       CNN  S(fused_sharp) > S(sharp) -0.519513  0.684587
7   TSS       CNN  S(fused_sharp) > S(sharp) -0.821323  0.771203
8   HSS       CNN  S(fused_sharp) > S(sharp) -0.821322  0.771203
9   BSS       CNN  S(fused_sharp) > S(sharp) -0.917732  0.794656
10  ACC      LSTM  S(fused_smarp) > S(smarp)  1.102187  0.166119
11  AUC      LSTM  S(fused_smarp) > S(smarp)  1.497672  0.104286
12  TSS      LSTM  S(fused_smarp) > S(smarp)  1.102190  0.166118
13  HSS      LSTM  S(fused_smarp) > S(smarp)  1.102182  0.166120
14  BSS      LSTM  S(fused_smarp) > S(smarp)  2.974497  0.020481
15  ACC       CNN  S(fused_smarp) > S(smarp)  3.181951  0.016736
16  AUC       CNN  S(fused_smarp) > S(smarp)  2.959646  0.020785
17  TSS       CNN  S(fused_smarp) > S(smarp)  3.181951  0.016736
18  HSS       CNN  S(fused_smarp) > S(smarp)  3.181953  0.016736
19  BSS       CNN  S(fused_smarp) > S(smarp)  3.928356  0.008563

In [39]:
df_ttest = (df_ttest
 .set_index(['S', 'estimator', 'tested hypothesis'])
 .sort_index() # group multiindex
 .unstack(-1)
 .swaplevel(axis=1)
 .sort_index(level=0, axis=1) # group column multiindex
).loc[['ACC', 'AUC', 'TSS', 'HSS', 'BSS']]

In [40]:
print(typeset(df_ttest))

\begin{tabular}{llrrrr}
\toprule
    & tested hypothesis & \multicolumn{2}{c}{S(fused\_sharp) > S(sharp)} & \multicolumn{2}{c}{S(fused\_smarp) > S(smarp)} \\
    & {} &                   p-value &         t &                   p-value &         t \\
S & estimator &                           &           &                           &           \\
\midrule
ACC & CNN &                  0.771203 & -0.821323 &                  0.016736 &  3.181951 \\
    & LSTM &                  0.027282 &  2.691665 &                  0.166119 &  1.102187 \\
AUC & CNN &                  0.684587 & -0.519513 &                  0.020785 &  2.959646 \\
    & LSTM &                  0.027924 &  2.669280 &                  0.104286 &  1.497672 \\
TSS & CNN &                  0.771203 & -0.821323 &                  0.016736 &  3.181951 \\
    & LSTM &                  0.027282 &  2.691671 &                  0.166118 &  1.102190 \\
HSS & CNN &                  0.771203 & -0.821322 &                  0.016736 &  3.

In [24]:
print_pvalues(runs_1, 'sharp')
print_pvalues(runs_1, 'smarp')

Is fused_sharp better than sharp?
LSTM
ACC (2.6916645246206277, 0.027281877684655762)
AUC (2.6692798026776545, 0.02792392028931603)
TSS (2.6916707765182757, 0.027281700760456128)
HSS (2.6916717584699468, 0.02728167297205153)
BSS (2.3511157478826203, 0.03920985724955582)
CNN
ACC (-0.8213233611662126, 0.7712032049868179)
AUC (-0.5195128291223914, 0.6845868990533817)
TSS (-0.8213225138367546, 0.7712029897685833)
HSS (-0.8213222599641519, 0.7712029252859439)
BSS (-0.9177320549958328, 0.794656009059732)
Is fused_smarp better than smarp?
LSTM
ACC (1.1021867887766423, 0.1661187740852072)
AUC (1.4976723753080161, 0.10428641828251642)
TSS (1.1021895363546819, 0.16611824316013168)
HSS (1.1021819898982497, 0.16611970139463986)
BSS (2.9744968497645057, 0.020480701244647126)
CNN
ACC (3.181950561266245, 0.016736351853538545)
AUC (2.959645758546297, 0.02078452572379359)
TSS (3.1819514737747063, 0.01673633724988792)
HSS (3.181952763462706, 0.01673631660993727)
BSS (3.9283561398381366, 0.00856334650828

In [8]:
runs_raw_2 = retrieve('leaderboard3', 'val_tss_2')
runs_2 = select(runs_raw_2)
runs_2.style.background_gradient(axis=0)

Select iloc slice(None, None, None) from 
                          start_time tags.mlflow.runName
204 2021-08-23 18:58:17.342000+00:00           val_tss_2


In [11]:
organize(runs_2, std=True)
# style(runs_2)

dataset          fused_sharp    fused_smarp          sharp          smarp
    estimator                                                            
ACC CNN        0.916+/-0.020  0.886+/-0.034  0.929+/-0.015  0.871+/-0.038
    LSTM       0.950+/-0.011  0.896+/-0.033  0.947+/-0.014  0.891+/-0.031
AUC CNN        0.982+/-0.005  0.955+/-0.021  0.981+/-0.006  0.945+/-0.025
    LSTM       0.991+/-0.004  0.961+/-0.019  0.987+/-0.004  0.958+/-0.019
TSS CNN        0.832+/-0.040  0.772+/-0.069  0.859+/-0.030  0.742+/-0.076
    LSTM       0.900+/-0.023  0.793+/-0.066  0.894+/-0.027  0.782+/-0.062
HSS CNN        0.832+/-0.040  0.772+/-0.069  0.859+/-0.030  0.742+/-0.076
    LSTM       0.900+/-0.023  0.793+/-0.066  0.894+/-0.027  0.782+/-0.062
BSS CNN        0.690+/-0.093  0.563+/-0.133  0.738+/-0.056  0.485+/-0.143
    LSTM       0.804+/-0.036  0.590+/-0.140  0.796+/-0.042  0.546+/-0.131

In [25]:
print_pvalues(runs_2, 'sharp')
print_pvalues(runs_2, 'smarp')

Is fused_sharp better than sharp?
LSTM
ACC (0.9777278319319056, 0.19178487897680624)
AUC (4.300694664396739, 0.006319335398415607)
TSS (0.9777315657744146, 0.1917840595337562)
HSS (0.977724850026493, 0.19178553339909488)
BSS (0.37855958867666417, 0.3621249216050012)
CNN
ACC (-0.9748417724434105, 0.8075808325498115)
AUC (0.23214946586334767, 0.4139077658595507)
TSS (-0.9748406612178896, 0.8075805879808844)
HSS (-0.9748411065632725, 0.8075806859967023)
BSS (-0.9058617234442904, 0.7918796982251726)
Is fused_smarp better than smarp?
LSTM
ACC (2.16984934972141, 0.04791313831179759)
AUC (1.3237969359757362, 0.12807811097155478)
TSS (2.1698549758415853, 0.047912836621057026)
HSS (2.1698511692452986, 0.04791304074281682)
BSS (6.5718264634542205, 0.0013872198615761167)
CNN
ACC (2.873014380272446, 0.022665371044264505)
AUC (3.3825810827091036, 0.01385962635772741)
TSS (2.8730112352296087, 0.022665442839425058)
HSS (2.873014380272446, 0.022665371044264505)
BSS (3.187187555462433, 0.01665278638005

In [10]:
runs_comb = select(pd.concat((runs_raw_1, runs_raw_2)))
(runs_comb
 .sort_values(['dataset', 'estimator', 'seed'])
 .style.background_gradient(axis=0)
)

#organize(runs_comb, std=True)

In [6]:
df = organize(runs_comb, std=True)
column_order = ['fused_sharp', 'sharp', 'fused_smarp', 'smarp']
df = (df
      .drop(index='HSS')
      .loc[:, column_order]
     )
df

/home/zeyusun/work/flare-prediction-smarp/notebooks/mlflow_helper.py:98: FutureWarning: Dropping invalid columns in DataFrameGroupBy.agg is deprecated. In a future version, a TypeError will be raised. Before calling .agg, select only columns which should be valid for the aggregating function.
  df = (runs


dataset          fused_sharp          sharp    fused_smarp          smarp
    estimator                                                            
ACC CNN        0.906+/-0.036  0.922+/-0.017  0.901+/-0.028  0.877+/-0.031
    LSTM       0.950+/-0.012  0.942+/-0.016  0.905+/-0.025  0.900+/-0.024
AUC CNN        0.980+/-0.009  0.981+/-0.006  0.963+/-0.017  0.950+/-0.020
    LSTM       0.990+/-0.004  0.986+/-0.004  0.966+/-0.015  0.963+/-0.015
TSS CNN        0.812+/-0.071  0.843+/-0.034  0.802+/-0.056  0.754+/-0.061
    LSTM       0.900+/-0.023  0.884+/-0.032  0.810+/-0.050  0.800+/-0.049
BSS CNN        0.649+/-0.152  0.714+/-0.064  0.628+/-0.114  0.520+/-0.121
    LSTM       0.799+/-0.036  0.775+/-0.047  0.626+/-0.107  0.586+/-0.108

In [7]:
print(typeset(df))

\begin{tabular}{llllll}
\toprule
    & dataset &    fused\_sharp &          sharp &    fused\_smarp &          smarp \\
{} & estimator &                &                &                &                \\
\midrule
\multirow{2}{*}{ACC} & CNN &  0.906+/-0.036 &  0.922+/-0.017 &  0.901+/-0.028 &  0.877+/-0.031 \\
    & LSTM &  0.950+/-0.012 &  0.942+/-0.016 &  0.905+/-0.025 &  0.900+/-0.024 \\
\cline{1-6}
\multirow{2}{*}{AUC} & CNN &  0.980+/-0.009 &  0.981+/-0.006 &  0.963+/-0.017 &  0.950+/-0.020 \\
    & LSTM &  0.990+/-0.004 &  0.986+/-0.004 &  0.966+/-0.015 &  0.963+/-0.015 \\
\cline{1-6}
\multirow{2}{*}{TSS} & CNN &  0.812+/-0.071 &  0.843+/-0.034 &  0.802+/-0.056 &  0.754+/-0.061 \\
    & LSTM &  0.900+/-0.023 &  0.884+/-0.032 &  0.810+/-0.050 &  0.800+/-0.049 \\
\cline{1-6}
\multirow{2}{*}{BSS} & CNN &  0.649+/-0.152 &  0.714+/-0.064 &  0.628+/-0.114 &  0.520+/-0.121 \\
    & LSTM &  0.799+/-0.036 &  0.775+/-0.047 &  0.626+/-0.107 &  0.586+/-0.108 \\
\bottomrule
\end{tabular}



In [8]:
# Just in case the order is messed up
print_pvalues(runs_comb.sort_values(by=['dataset', 'estimator', 'seed']), 'sharp')
print_pvalues(runs_comb.sort_values(by=['dataset', 'estimator', 'seed']), 'smarp')

Is fused_sharp better than sharp?
LSTM
ACC (2.5140740674884015, 0.016544411931847147)
AUC (4.842484590348467, 0.0004589266849649696)
TSS (2.514079223405149, 0.016544271968921782)
HSS (2.5140754830901475, 0.016544373503691832)
BSS (1.7750818450355275, 0.05481202007976016)
CNN
ACC (-1.2923586700281378, 0.8857867848637838)
AUC (-0.23388120441526536, 0.589844716114039)
TSS (-1.292357174923711, 0.8857865371818603)
HSS (-1.2923571562579335, 0.8857865340896481)
BSS (-1.3145833878238404, 0.8894192694044404)
Is fused_smarp better than smarp?
LSTM
ACC (2.219666212347629, 0.02679661722853573)
AUC (2.0745721253722595, 0.03392974779747737)
TSS (2.2196730868047756, 0.026796316590263016)
HSS (2.219661749765017, 0.026796812390891222)
BSS (6.014783836597528, 9.943190037371047e-05)
CNN
ACC (3.8811365757125365, 0.0018624708666261554)
AUC (4.070351511021532, 0.0013990854701769534)
TSS (3.8811352512795105, 0.0018624746279870177)
HSS (3.8811369866243663, 0.001862469699647167)
BSS (4.806836601989046, 0.00048

In [11]:
df_ttest = tabulate_pvalues(runs_comb)

In [16]:
df_ttest_print.index

MultiIndex([('ACC',  'CNN'),
            ('ACC', 'LSTM'),
            ('AUC',  'CNN'),
            ('AUC', 'LSTM'),
            ('TSS',  'CNN'),
            ('TSS', 'LSTM'),
            ('BSS',  'CNN'),
            ('BSS', 'LSTM')],
           names=['S', 'estimator'])

In [21]:
df_ttest_print = (
    df_ttest
    .set_index(['S', 'estimator', 'tested hypothesis'])
    .sort_index() # group multiindex
    .unstack(-1)
    .swaplevel(axis=1)
    .sort_index(level=0, axis=1) # group column multiindex
    .loc[['ACC', 'AUC', 'TSS', 'BSS']] # sort index
    .rename(columns={
        'S(fused_sharp) > S(sharp)': '$S_{\texttt{FUSED\_SHARP}} > S_{\texttt{SHARP\_ONLY}}$',
        'S(fused_smarp) > S(smarp)': '$S_{\texttt{FUSED\_SMARP}} > S_{\texttt{SMARP\_ONLY}}$',
        'p-value': '$p$-value',
        't': '$t$',
    })
    .rename_axis(
        index=['Metric $S$', 'Estimator'],
        columns=['$H_1$', ''],
    )
)
df_ttest_print

$H_1$                $S_{\texttt{FUSED\_SHARP}} > S_{\texttt{SHARP\_ONLY}}$  \
                                                                  $p$-value   
Metric $S$ Estimator                                                          
ACC        CNN                                                     0.885787   
           LSTM                                                    0.016544   
AUC        CNN                                                     0.589845   
           LSTM                                                    0.000459   
TSS        CNN                                                     0.885787   
           LSTM                                                    0.016544   
BSS        CNN                                                     0.889419   
           LSTM                                                    0.054812   

$H_1$                           \
                           $t$   
Metric $S$ Estimator             
ACC        CNN       -1.292359   
           LSTM       2.514074   
AUC        CNN       -0.233881   
           LSTM       4.842485   
TSS        CNN       -1.292357   
           LSTM       2.514079   
BSS        CNN       -1.314583   
           LSTM       1.775082   

$H_1$                $S_{\texttt{FUSED\_SMARP}} > S_{\texttt{SMARP\_ONLY}}$  \
                                                                  $p$-value   
Metric $S$ Estimator                                                          
ACC        CNN                                                     0.001862   
           LSTM                                                    0.026797   
AUC        CNN                                                     0.001399   
           LSTM                                                    0.033930   
TSS        CNN                                                     0.001862   
           LSTM                                                    0.026796   
BSS        CNN                                                     0.000482   
           LSTM                                                    0.000099   

$H_1$                           
                           $t$  
Metric $S$ Estimator            
ACC        CNN        3.881137  
           LSTM       2.219666  
AUC        CNN        4.070352  
           LSTM       2.074572  
TSS        CNN        3.881135  
           LSTM       2.219673  
BSS        CNN        4.806837  
           LSTM       6.014784

In [30]:
print(typeset(df_ttest_print, escape=False))

\begin{tabular}{llrrrr}
\toprule
    & $H_1$ & \multicolumn{2}{c}{$S_{\texttt{FUSED\_SHARP}} > S_{\texttt{SHARP\_ONLY}}$} & \multicolumn{2}{c}{$S_{\texttt{FUSED\_SMARP}} > S_{\texttt{SMARP\_ONLY}}$} \\
    & {} &                                              $p$-value &       $t$ &                                              $p$-value &       $t$ \\
Metric $S$ & Estimator &                                                        &           &                                                        &           \\
\midrule
\multirow{2}{*}{ACC} & CNN &                                               0.885787 & -1.292359 &                                               0.001862 &  3.881137 \\
    & LSTM &                                               0.016544 &  2.514074 &                                               0.026797 &  2.219666 \\
\cline{1-6}
\multirow{2}{*}{AUC} & CNN &                                               0.589845 & -0.233881 &                                              

In [5]:
runs_raw_3 = retrieve('CNN', 'val_tss_CNN_larger_lr')
runs_3 = select(runs_raw_3)
runs_3.style.background_gradient(axis=0)

Select iloc 0 from 
                         start_time    tags.mlflow.runName  \
20 2021-08-23 19:05:50.497000+00:00  val_tss_CNN_larger_lr   

               tags.mlflow.source.git.commit  
20  cb46e19c92c4ff4b078a69556524e23ef1282642  


In [29]:
organize(runs_3, std=True)

dataset          fused_sharp          sharp
    estimator                              
ACC CNN        0.904+/-0.036  0.915+/-0.019
AUC CNN        0.977+/-0.012  0.983+/-0.006
TSS CNN        0.809+/-0.073  0.831+/-0.037
HSS CNN        0.809+/-0.073  0.831+/-0.037
BSS CNN        0.648+/-0.131  0.698+/-0.072

In [30]:
# We observe no significant difference increasing the learning rate from 1e-3 to 1e-2
diff(
    runs_comb[
        (runs_comb['estimator'] == 'CNN') &
        (runs_comb['dataset'].isin(['fused_sharp', 'sharp']))
    ].sort_values(['dataset', 'seed']),
    runs_3.sort_values(['dataset', 'seed'])
).style.background_gradient(axis=0, vmin=-0.15, vmax=0.15, cmap='seismic')

## LSTM: lstm_more_patience
No significant increase. Some even decrease.

## ensemble: disjoint_train

In [36]:
runs_raw = retrieve('ensemble', 'disjoint_train')
runs = select(runs_raw)
runs.style.background_gradient(axis=0)

Select iloc 0 from 
                         start_time tags.mlflow.runName  \
40 2021-08-26 06:51:54.064000+00:00      disjoint_train   

               tags.mlflow.source.git.commit  
40  6cca85c0c886f98b13533dfbbb5bd8c832c43dff  


In [37]:
organize(runs, std=True)

dataset          fused_sharp    fused_smarp          sharp          smarp
    estimator                                                            
ACC CNN        0.863+/-0.013  0.899+/-0.016  0.918+/-0.025  0.891+/-0.024
    LSTM       0.946+/-0.012  0.909+/-0.012  0.936+/-0.018  0.909+/-0.013
AUC CNN        0.965+/-0.016  0.965+/-0.003  0.983+/-0.005  0.957+/-0.013
    LSTM       0.990+/-0.004  0.967+/-0.010  0.985+/-0.004  0.968+/-0.011
TSS CNN        0.726+/-0.026  0.797+/-0.032  0.835+/-0.050  0.782+/-0.047
    LSTM       0.893+/-0.024  0.818+/-0.024  0.873+/-0.036  0.818+/-0.026
HSS CNN        0.726+/-0.026  0.797+/-0.032  0.835+/-0.050  0.782+/-0.047
    LSTM       0.893+/-0.024  0.818+/-0.024  0.873+/-0.036  0.818+/-0.026
BSS CNN        0.486+/-0.105  0.621+/-0.050  0.712+/-0.090  0.587+/-0.079
    LSTM       0.772+/-0.056  0.651+/-0.060  0.755+/-0.045  0.626+/-0.072

In [38]:
print_pvalues(runs, 'sharp')
print_pvalues(runs, 'smarp')

Is fused_sharp better than sharp?
LSTM
ACC (2.1688139857412985, 0.04796869389184776)
AUC (3.252673981342178, 0.01564827559870736)
TSS (2.1688142590789883, 0.04796867921562959)
HSS (2.1688194750469654, 0.04796839915757291)
BSS (1.2634251930993408, 0.13752878220045292)
CNN
ACC (-6.071624291713522, 0.9981414333885852)
AUC (-3.0368820350165717, 0.9807401377742772)
TSS (-6.071620703092071, 0.9981414293545399)
HSS (-6.071624896191048, 0.9981414340680901)
BSS (-8.153401677201131, 0.9993842257101654)
Is fused_smarp better than smarp?
LSTM
ACC (-0.15195033824737603, 0.5567089366845407)
AUC (-0.47752877857065323, 0.671045969380375)
TSS (-0.15195471745342326, 0.5567105554255567)
HSS (-0.1519546760822291, 0.556710540133008)
BSS (1.8711467233535082, 0.06732619857231695)
CNN
ACC (0.7192249113784385, 0.255890184537878)
AUC (1.3540030085999257, 0.12358964105220792)
TSS (0.7192273188562722, 0.2558895184224)
HSS (0.7192253269654988, 0.25589006955065385)
BSS (0.9598334047801804, 0.19574679666721567)


## Revisit leaderboard3: LSTM_CNN

In [40]:
runs_raw = retrieve('leaderboard3', 'LSTM_CNN')
runs = select(runs_raw)
runs.style.background_gradient(axis=0)

Select iloc 0 from 
                          start_time tags.mlflow.runName  \
289 2021-06-27 16:03:33.239000+00:00            LSTM_CNN   

                tags.mlflow.source.git.commit  
289  e13810ffddafdb818df7c4de68d9eafd67861e8c  


In [42]:
organize(runs, std=True)

dataset          fused_sharp    fused_smarp          sharp          smarp
    estimator                                                            
ACC CNN        0.943+/-0.030  0.910+/-0.019  0.934+/-0.011  0.888+/-0.021
    LSTM       0.957+/-0.012  0.921+/-0.016  0.951+/-0.012  0.918+/-0.017
AUC CNN        0.983+/-0.013  0.964+/-0.012  0.981+/-0.004  0.956+/-0.016
    LSTM       0.988+/-0.004  0.970+/-0.010  0.986+/-0.005  0.965+/-0.009
TSS CNN        0.886+/-0.059  0.820+/-0.038  0.867+/-0.023  0.776+/-0.041
    LSTM       0.913+/-0.025  0.841+/-0.032  0.901+/-0.024  0.836+/-0.034
HSS CNN        0.886+/-0.059  0.820+/-0.038  0.867+/-0.023  0.776+/-0.041
    LSTM       0.913+/-0.025  0.841+/-0.032  0.901+/-0.024  0.836+/-0.034
BSS CNN        0.343+/-0.388  0.365+/-0.305  0.619+/-0.125  0.274+/-0.493
    LSTM       0.793+/-0.055  0.621+/-0.056  0.757+/-0.062  0.581+/-0.047

In [43]:
print_pvalues(runs, 'sharp')
print_pvalues(runs, 'smarp')

Is fused_sharp better than sharp?
LSTM
ACC (1.689632944420116, 0.08318224343705961)
AUC (2.0214023658182074, 0.056658282064086796)
TSS (1.6896246902065741, 0.0831830485610414)
HSS (1.689625557197938, 0.08318296399347924)
BSS (3.044136124285647, 0.01912351051340325)
CNN
ACC (0.7567072805389897, 0.24567453340671816)
AUC (0.25988312670010305, 0.40389122223273266)
TSS (0.7567063288785028, 0.24567478882620117)
HSS (0.7567058351632051, 0.24567492133625438)
BSS (-1.4145173522604462, 0.8849415126274763)
Is fused_smarp better than smarp?
LSTM
ACC (3.0337646190777083, 0.019318808620029243)
AUC (5.564262767009857, 0.002554576351214866)
TSS (3.033800086302704, 0.019318136791867425)
HSS (3.0337754398937893, 0.019318603646792762)
BSS (3.6452319105298883, 0.010931663760441877)
CNN
ACC (2.3060256013512648, 0.04119566155262506)
AUC (2.1449907642195116, 0.04926700349103811)
TSS (2.3060276185231734, 0.04119557021588522)
HSS (2.3060276185231734, 0.04119557021588522)
BSS (0.5976277349558164, 0.291138343976

## ensemble: disjoint_train
Train on SMARP

In [4]:
runs_raw = retrieve('ensemble', 'disjoint_train')
runs = select(runs_raw)
runs.style.background_gradient(axis=0)

Select iloc 0 from 
                         start_time tags.mlflow.runName  \
40 2021-08-26 06:51:54.064000+00:00      disjoint_train   

               tags.mlflow.source.git.commit  
40  6cca85c0c886f98b13533dfbbb5bd8c832c43dff  


In [5]:
organize(runs, std=True)

dataset          fused_sharp    fused_smarp          sharp          smarp
    estimator                                                            
ACC CNN        0.863+/-0.013  0.899+/-0.016  0.918+/-0.025  0.891+/-0.024
    LSTM       0.946+/-0.012  0.909+/-0.012  0.936+/-0.018  0.909+/-0.013
AUC CNN        0.965+/-0.016  0.965+/-0.003  0.983+/-0.005  0.957+/-0.013
    LSTM       0.990+/-0.004  0.967+/-0.010  0.985+/-0.004  0.968+/-0.011
TSS CNN        0.726+/-0.026  0.797+/-0.032  0.835+/-0.050  0.782+/-0.047
    LSTM       0.893+/-0.024  0.818+/-0.024  0.873+/-0.036  0.818+/-0.026
HSS CNN        0.726+/-0.026  0.797+/-0.032  0.835+/-0.050  0.782+/-0.047
    LSTM       0.893+/-0.024  0.818+/-0.024  0.873+/-0.036  0.818+/-0.026
BSS CNN        0.486+/-0.105  0.621+/-0.050  0.712+/-0.090  0.587+/-0.079
    LSTM       0.772+/-0.056  0.651+/-0.060  0.755+/-0.045  0.626+/-0.072

LSTM did better on `fused_sharp` than on `sharp`, and almost the same on `fused_smarp` and `smarp`.

CNN did worse on `fused_sharp` than on `sharp`, and  better on `fused_smarp` than `smarp`.

In [6]:
print_pvalues(runs, 'sharp')
print_pvalues(runs, 'smarp')

Is fused_sharp better than sharp?
LSTM
ACC (2.1688139857412985, 0.04796869389184776)
AUC (3.252673981342178, 0.01564827559870736)
TSS (2.1688142590789883, 0.04796867921562959)
HSS (2.1688194750469654, 0.04796839915757291)
BSS (1.2634251930993408, 0.13752878220045292)
CNN
ACC (-6.071624291713522, 0.9981414333885852)
AUC (-3.0368820350165717, 0.9807401377742772)
TSS (-6.071620703092071, 0.9981414293545399)
HSS (-6.071624896191048, 0.9981414340680901)
BSS (-8.153401677201131, 0.9993842257101654)
Is fused_smarp better than smarp?
LSTM
ACC (-0.15195033824737603, 0.5567089366845407)
AUC (-0.47752877857065323, 0.671045969380375)
TSS (-0.15195471745342326, 0.5567105554255567)
HSS (-0.1519546760822291, 0.556710540133008)
BSS (1.8711467233535082, 0.06732619857231695)
CNN
ACC (0.7192249113784385, 0.255890184537878)
AUC (1.3540030085999257, 0.12358964105220792)
TSS (0.7192273188562722, 0.2558895184224)
HSS (0.7192253269654988, 0.25589006955065385)
BSS (0.9598334047801804, 0.19574679666721567)


In [13]:
tabulate_pvalues(runs)

S      dataset tested hypothesis         t   p-value
0   ACC  fused_sharp  S(LSTM) > S(CNN)  5.246785  0.000265
1   AUC  fused_sharp  S(LSTM) > S(CNN)  4.728813  0.000538
2   TSS  fused_sharp  S(LSTM) > S(CNN)  5.246787  0.000265
3   HSS  fused_sharp  S(LSTM) > S(CNN)  5.246786  0.000265
4   BSS  fused_sharp  S(LSTM) > S(CNN)  5.259573  0.000261
5   ACC        sharp  S(LSTM) > S(CNN)  3.209300  0.005335
6   AUC        sharp  S(LSTM) > S(CNN)  2.457645  0.018149
7   TSS        sharp  S(LSTM) > S(CNN)  3.209302  0.005335
8   HSS        sharp  S(LSTM) > S(CNN)  3.209304  0.005335
9   BSS        sharp  S(LSTM) > S(CNN)  3.438200  0.003706
10  ACC  fused_smarp  S(LSTM) > S(CNN)  0.487082  0.318922
11  AUC  fused_smarp  S(LSTM) > S(CNN)  1.288573  0.114842
12  TSS  fused_smarp  S(LSTM) > S(CNN)  0.487081  0.318922
13  HSS  fused_smarp  S(LSTM) > S(CNN)  0.487081  0.318922
14  BSS  fused_smarp  S(LSTM) > S(CNN) -0.218546  0.584060
15  ACC        smarp  S(LSTM) > S(CNN)  3.269905  0.004842
16  AUC        smarp  S(LSTM) > S(CNN)  4.097666  0.001343
17  TSS        smarp  S(LSTM) > S(CNN)  3.269906  0.004842
18  HSS        smarp  S(LSTM) > S(CNN)  3.269906  0.004842
19  BSS        smarp  S(LSTM) > S(CNN)  2.444970  0.018531

## leaderboard8: reproduce

In [50]:
runs_raw = retrieve('leaderboard8', 'reproduce')
runs = select(runs_raw)
(runs
 .loc[(runs['estimator'] == 'LSTM') & (runs['dataset'].isin(['fused_smarp', 'smarp']))]
 .loc[:, ['seed', 'dataset', 'ACC', 'AUC', 'TSS', 'BSS']]
 .set_index(['seed', 'dataset'])
 .sort_values(['seed', 'dataset'])
 #.sort_values(['dataset', 'estimator', 'seed'])
 #[(runs['seed'] >= '0') & (runs['seed'] <= '4')]
 .style.background_gradient(axis=0)
)

In [4]:
organize(runs, std=True)

/home/zeyusun/work/flare-prediction-smarp/notebooks/mlflow_helper.py:98: FutureWarning: Dropping invalid columns in DataFrameGroupBy.agg is deprecated. In a future version, a TypeError will be raised. Before calling .agg, select only columns which should be valid for the aggregating function.
  df = (runs


dataset          fused_sharp    fused_smarp          sharp          smarp
    estimator                                                            
ACC CNN        0.907+/-0.029  0.897+/-0.025  0.913+/-0.027  0.876+/-0.030
    LSTM       0.952+/-0.013  0.900+/-0.025  0.939+/-0.021  0.900+/-0.025
AUC CNN        0.977+/-0.010  0.961+/-0.016  0.975+/-0.018  0.949+/-0.020
    LSTM       0.990+/-0.004  0.965+/-0.015  0.987+/-0.005  0.964+/-0.015
TSS CNN        0.815+/-0.057  0.794+/-0.051  0.827+/-0.055  0.753+/-0.060
    LSTM       0.904+/-0.027  0.801+/-0.051  0.878+/-0.042  0.800+/-0.049
HSS CNN        0.815+/-0.057  0.794+/-0.051  0.827+/-0.055  0.753+/-0.060
    LSTM       0.904+/-0.027  0.801+/-0.051  0.878+/-0.042  0.800+/-0.049
BSS CNN        0.636+/-0.123  0.610+/-0.104  0.681+/-0.109  0.514+/-0.134
    LSTM       0.806+/-0.043  0.604+/-0.110  0.772+/-0.065  0.576+/-0.096

In [5]:
print_pvalues(runs, 'sharp')
print_pvalues(runs, 'smarp')

Is fused_sharp better than sharp?
LSTM
ACC (3.690383389739849, 0.0024972080200710524)
AUC (4.990945716712597, 0.00037408057077634394)
TSS (3.6903891198597965, 0.0024971858439906632)
HSS (3.6903845603234298, 0.0024972034897902933)
BSS (1.9939220480241295, 0.03865325025224283)
CNN
ACC (-0.8102409924271193, 0.780650157137726)
AUC (0.474529751650027, 0.3232123108264322)
TSS (-0.8102397916474471, 0.7806498294574424)
HSS (-0.810240466889097, 0.7806500137239063)
BSS (-1.6383121476392117, 0.9321097111797362)
Is fused_smarp better than smarp?
LSTM
ACC (0.17071653453069832, 0.4341114751766456)
AUC (1.3889643454548641, 0.09912453295655363)
TSS (0.17070528560043458, 0.43411577016610203)
HSS (0.17071173134780812, 0.43411330909389967)
BSS (2.252679914459496, 0.025390210974998183)
CNN
ACC (1.9042476073798495, 0.044639771921050085)
AUC (3.6556268072823133, 0.0026356092860227286)
TSS (1.9042475162761787, 0.04463977842628509)
HSS (1.9042474335817572, 0.0446397843310605)
BSS (2.4261370613241655, 0.019111

In [42]:
df_ttest = tabulate_pvalues(runs)
df_ttest = (df_ttest
            .loc[df_ttest['estimator'] == 'LSTM']
 .set_index(['S', 'estimator', 'tested hypothesis'])
 .sort_index() # group multiindex
 .unstack(-1)
 .swaplevel(axis=1)
 .sort_index(level=0, axis=1) # group column multiindex
).loc[['ACC', 'AUC', 'TSS', 'BSS']]
df_ttest.style.applymap(
    lambda x: 'background-color : yellow' if x<0.05 else '',
    subset=(slice(None), [True, False, True, False]),
)

* For LSTM, data fusion helps
    * when adding SMARP in training and testing on SHARP
    * but not when adding SHARP in training and testing on SMARP
* Conclusion 1: Does that mean SMARP keywords has more information related to flares?
* For CNN, data fusion helps
    * when adding SHARP magnetograms and testing on SMARP
    * but not when adding SMARP magnetograms and testing on SHARP
* Conclusion 2: Does that suggest SHARP magnetograms has more information related to flares?
* Conclusion 1 & 2 are in conflict.
    * The difference may not be due to how well data itself is related to flares. It could also be the way we fuse the data. But if there is something wrong, for parameters. E.g., it could be that when adding SHARP parameters to the training set, we do not normalize correctly, but when the other way around, we normalized SMARP correctly. We did normalizaiton separately for SHARP and SMARP. 
* Inspection: Check the distributions *after* transformation.
    * But both SMARP and SHARP should have mean 0 and std 1 after standardization.
    * Does it make sense to normalization the two datasets separately?
        * One cycle could have fewer flares, and parameters are smaller. Normalizing separately artificially increase the activity of this cycle.
        * Normalizing separately assumes that the data for different solar cycles have the same mean and same std for the "true value" of every keyword.
            * The keyword parameters are measurements of the "true value". It may also depends on the unit.
            * If the first two moments are different, then separated normalization will create a bias to one of the dataset, making its values larger or more variable.
        * RUS doesn't make a difference. We are not computing mean and std with RUS-ed data.
        * The problem is, we are not able to recover the "true value".
            * We can derive the transformation using the co-temporal measurements. This should generalize well.
            * We can convert MEANGBL to the same physical unit. But this should be accounted for in the linear transformation derived above.
            * This seems a valid approach. Why we decided not to use it?
                * Because we decided to normalize separately. In that case, performing a transformation to one of them will have no effect after normalization of that dataset.
                * However, the transformation certainly makes a difference if we perform a global normalization. Indeed, the normalization should be performed after correction/calibration/fusion. You have messed up the order.